Raise `BufferError` for DLPack buffer-related errors.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150691
* #150218
* #150217
* #150216

This PR addresses the Array API documentation for [`__dlpack__`][1] and
[`from_dlpack`][2] by making some buffer-related errors `BufferError`
instead of `RuntimeError`, e.g. incompatible dtype, strides, or device.

[1]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html
[2]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.from_dlpack.html#from-dlpack

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

msaroufim commented on aten/src/ATen/DLConvertor.cpp:
```diff
@@ -69,29 +69,29 @@ DLDataType getDLDataType(const Tensor& t) {
     case ScalarType::Float8_e4m3fnuz:
     case ScalarType::Float8_e8m0fnu:
-      TORCH_CHECK(false, "float8 types are not supported by dlpack");
+      TORCH_CHECK_BUFFER(false, "float8 types are not supported by dlpack");
```
maybe the more specific error would be not supported by 1.0, it is supported for example by 1.1 now https://github.com/dmlc/dlpack/releases

ysiraichi replied:
Sure we could do that. But, maybe in a later PR.

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Rebased.

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update on "Raise `BufferError` for DLPack buffer-related errors."


This PR addresses the Array API documentation for [`__dlpack__`][1] and
[`from_dlpack`][2] by making some buffer-related errors `BufferError`
instead of `RuntimeError`, e.g. incompatible dtype, strides, or device.

[1]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html
[2]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.from_dlpack.html#from-dlpack

[ghstack-poisoned]

ysiraichi committed:
Update on "Raise `BufferError` for DLPack buffer-related errors."


This PR addresses the Array API documentation for [`__dlpack__`][1] and
[`from_dlpack`][2] by making some buffer-related errors `BufferError`
instead of `RuntimeError`, e.g. incompatible dtype, strides, or device.

[1]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html
[2]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.from_dlpack.html#from-dlpack

[ghstack-poisoned]

ysiraichi commented:
@albanD I think I have addressed all of the reviews.
Could you take a look at this stack of PRs, again?

ysiraichi committed:
Update on "Raise `BufferError` for DLPack buffer-related errors."


This PR addresses the Array API documentation for [`__dlpack__`][1] and
[`from_dlpack`][2] by making some buffer-related errors `BufferError`
instead of `RuntimeError`, e.g. incompatible dtype, strides, or device.

[1]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html
[2]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.from_dlpack.html#from-dlpack

[ghstack-poisoned]

ysiraichi committed:
Update on "Raise `BufferError` for DLPack buffer-related errors."


This PR addresses the Array API documentation for [`__dlpack__`][1] and
[`from_dlpack`][2] by making some buffer-related errors `BufferError`
instead of `RuntimeError`, e.g. incompatible dtype, strides, or device.

[1]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html
[2]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.from_dlpack.html#from-dlpack

[ghstack-poisoned]

ysiraichi committed:
Update on "Raise `BufferError` for DLPack buffer-related errors."


This PR addresses the Array API documentation for [`__dlpack__`][1] and
[`from_dlpack`][2] by making some buffer-related errors `BufferError`
instead of `RuntimeError`, e.g. incompatible dtype, strides, or device.

[1]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html
[2]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.from_dlpack.html#from-dlpack

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge

ysiraichi committed:
Rebased. on "Raise `BufferError` for DLPack buffer-related errors."


This PR addresses the Array API documentation for [`__dlpack__`][1] and
[`from_dlpack`][2] by making some buffer-related errors `BufferError`
instead of `RuntimeError`, e.g. incompatible dtype, strides, or device.

[1]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html
[2]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.from_dlpack.html#from-dlpack

[ghstack-poisoned]

ysiraichi committed:
Update on "Raise `BufferError` for DLPack buffer-related errors."


This PR addresses the Array API documentation for [`__dlpack__`][1] and
[`from_dlpack`][2] by making some buffer-related errors `BufferError`
instead of `RuntimeError`, e.g. incompatible dtype, strides, or device.

[1]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html
[2]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.from_dlpack.html#from-dlpack

[ghstack-poisoned]

ysiraichi committed:
Update on "Raise `BufferError` for DLPack buffer-related errors."


This PR addresses the Array API documentation for [`__dlpack__`][1] and
[`from_dlpack`][2] by making some buffer-related errors `BufferError`
instead of `RuntimeError`, e.g. incompatible dtype, strides, or device.

[1]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html
[2]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.from_dlpack.html#from-dlpack

[ghstack-poisoned]

ysiraichi committed:
Update on "Raise `BufferError` for DLPack buffer-related errors."


This PR addresses the Array API documentation for [`__dlpack__`][1] and
[`from_dlpack`][2] by making some buffer-related errors `BufferError`
instead of `RuntimeError`, e.g. incompatible dtype, strides, or device.

[1]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html
[2]: https://data-apis.org/array-api/latest/API_specification/generated/array_api.from_dlpack.html#from-dlpack

[ghstack-poisoned]
